core/shared/src/main/scala/zio/NonEmptyChunk.scala

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import scala.language.implicitConversions
+
+import zio.NonEmptyChunk._
+
+/**
+ * A `NonEmptyChunk` is a `Chunk` that is guaranteed to contain at least one
+ * element. As a result, operations which would not be safe when performed on
+ * `Chunk`, such as `head` or `reduce`, are safe when performed on
+ * `NonEmptyChunk`. Operations on `NonEmptyChunk` which could potentially
+ * return an empty chunk will return a `Chunk` instead.
+ */
+final class NonEmptyChunk[+A] private (private val chunk: Chunk[A]) { self =>
```

Comment:
Could we make this a value class?

Reply:
Potentially. The issue there is that then it is the same class at runtime as `Chunk` so methods that have overloads for `Chunk` and `NonEmptyChunk` like `foreach` cause compiler errors. We could possibly resolve by adding a dummy parameter to them.